### PR TITLE
Add SHA-512 hardware acceleration on ARMv8

### PR DESCRIPTION
**Pull request by securitykernel**
Adds support for ARMv8 hardware acceleration using instrinsics. It is heavily based on the existing SHA-256 implementation with changes to use the equivalent intrinsics for SHA-512.

Tested on a MacBook Pro with M2 Pro.

Baseline software:

```bash
$ ./botan speed --msec=3000 SHA-512
SHA-512 hash buffer size 1024 bytes: 489.480 MiB/sec (1468.440 MiB in 3000.000 ms)
```

ARMv8 intrinsics:

```bash
$ ./botan speed --msec=3000 SHA-512
SHA-512 hash buffer size 1024 bytes: 1307.809 MiB/sec (3923.428 MiB in 3000.000 ms)
```

Which means a factor 2,67 or 267% performance increase.

Even beats OpenSSL on the same machine (via Homebrew):

```bash
$ src/scripts/bench.py SHA-512
   INFO: Comparing Botan 3.3.0 with OpenSSL 3.2.0
   INFO: Running OpenSSL benchmark for SHA-512
   INFO: Running Botan benchmark for SHA-512
algo SHA-512 buf_size     16 botan    158735807 bps openssl     84757824 bps adv botan by 1.87
algo SHA-512 buf_size     64 botan    756406746 bps openssl    340762560 bps adv botan by 2.22
algo SHA-512 buf_size    256 botan    950167132 bps openssl    743652864 bps adv botan by 1.28
algo SHA-512 buf_size   1024 botan   1386804433 bps openssl   1238273024 bps adv botan by 1.12
algo SHA-512 buf_size   8192 botan   1565492078 bps openssl   1518950906 bps adv botan by 1.03
algo SHA-512 buf_size  16384 botan   1579776447 bps openssl   1556676608 bps adv botan by 1.01
```

Before, the baseline software was significantly slower:

```bash
$ src/scripts/bench.py SHA-512
   INFO: Comparing Botan 3.3.0 with OpenSSL 3.2.0
   INFO: Running OpenSSL benchmark for SHA-512
   INFO: Running Botan benchmark for SHA-512
algo SHA-512 buf_size     16 botan     69516433 bps openssl     82078000 bps adv openssl by 1.18
algo SHA-512 buf_size     64 botan    289047132 bps openssl    318918273 bps adv openssl by 1.10
algo SHA-512 buf_size    256 botan    389636626 bps openssl    711016210 bps adv openssl by 1.82
algo SHA-512 buf_size   1024 botan    504381686 bps openssl   1153344512 bps adv openssl by 2.29
algo SHA-512 buf_size   8192 botan    563204317 bps openssl   1458631111 bps adv openssl by 2.59
algo SHA-512 buf_size  16384 botan    580899115 bps openssl   1547288576 bps adv openssl by 2.66
```

### CODE REVIEW

File: src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+      0x113F9804BEF90DAE, 0x1B710B35131C471B, 0x28DB77F523047D84, 0x32CAAB7B40C72493, 0x3C9EBE0A15C9BEBC,
+      0x431D67C49C100D4C, 0x4CC5D4BECB3E42B6, 0x597F299CFC657E2A, 0x5FCB6FAB3AD6FAEC, 0x6C44198C4A475817};
+
+   // Load initial values
+   uint64x2_t STATE0 = vld1q_u64(&digest[0]);  // ab
+   uint64x2_t STATE1 = vld1q_u64(&digest[2]);  // cd
+   uint64x2_t STATE2 = vld1q_u64(&digest[4]);  // ef
+   uint64x2_t STATE3 = vld1q_u64(&digest[6]);  // gh
+
+   // intermediate void* cast due to https://llvm.org/bugs/show_bug.cgi?id=20670
```
Comment (randombit) [inline src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:43]: This comment may no longer be relevant, can you try removing the cast and checking?

File: src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE0);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE3, STATE0, 1), vextq_u64(STATE2, STATE3, 1));
+      STATE0 = vsha512h2q_u64(TSTATE1, STATE2, STATE1);
+      STATE2 = vaddq_u64(STATE2, TSTATE1);
+      MSG7 = vsha512su1q_u64(vsha512su0q_u64(MSG7, MSG0), MSG6, vextq_u64(MSG3, MSG4, 1));
+
+      ////// UNROLLED LOOP //////
+
+      // TODOs:
+      // tests with long inputs fail
```
Comment (randombit) [inline src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:140]: ?!?

**Comment by randombit** [comment]
Regarding

```

algo SHA-512 buf_size     16 botan    158735807 bps openssl     84757824 bps adv botan by 1.87
algo SHA-512 buf_size     64 botan    756406746 bps openssl    340762560 bps adv botan by 2.22
algo SHA-512 buf_size    256 botan    950167132 bps openssl    743652864 bps adv botan by 1.28
algo SHA-512 buf_size   1024 botan   1386804433 bps openssl   1238273024 bps adv botan by 1.12
algo SHA-512 buf_size   8192 botan   1565492078 bps openssl   1518950906 bps adv botan by 1.03
algo SHA-512 buf_size  16384 botan   1579776447 bps openssl   1556676608 bps adv botan by 1.01
```

you probably noticed that our advantage is confined to the smaller input sizes. I recently had to benchmark several (Rust) SHA-256 libraries using SHA-NI and noticed something similar; basically every implementation was fast for long messages, but for small messages there were huge swings in performance, and with OpenSSL always the worst. I believe this is due to differences in buffer management. This is a good sign that we're doing the right thing in this regard.

**Review by randombit — APPROVED** [review]
Two small things but otherwise looks good to me, thanks!

File: src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp (hunk 2 above)
Comment (securitykernel) [reply src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:140]: Embarrassing leftover from debugging, removed it.

**Comment by randombit** [comment]
CI failures look relevent, but are confusing.

iOS:

```

  src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:76:16: error: always_inline function 'vsha512h2q_u64' requires target feature 'sha3', but would be inlined into function 'compress_digest_armv8' that is compiled without support for 'sha3'
        STATE3 = vsha512h2q_u64(TSTATE1, STATE1, STATE0);
```

sha3?

Android:

```

  src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:94:14: error: always_inline function 'vsha512su1q_u64' requires target feature 'sha3', but would be inlined into function 'compress_digest_armv8' that is compiled without support for 'sha3'
        MSG2 = vsha512su1q_u64(vsha512su0q_u64(MSG2, MSG3), MSG1, vextq_u64(MSG6, MSG7, 1));
```

sha3?

GCC

```
  In file included from src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:10:
  /usr/lib/gcc-cross/aarch64-linux-gnu/11/include/arm_neon.h:32278:1: error: inlining failed in call to ‘always_inline’ ‘uint64x2_t vsha512h2q_u64(uint64x2_t, uint64x2_t, uint64x2_t)’: target specific option mismatch
  32278 | vsha512h2q_u64 (uint64x2_t __a, uint64x2_t __b, uint64x2_t __c)
        | ^~~~~~~~~~~~~~
  src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp:374:30: note: called from here
    374 |       STATE1 = vsha512h2q_u64(TSTATE1, STATE3, STATE2);
        |                ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
```

Doesn't specify but I'm going to go ahead and guess it's sha3

**Comment by randombit** [comment]
Oh right, instead of `armv8crypto` ISA flag you need `armv8sha512`. For whatever reason ARM bundles the SHA-512 and SHA-3 instructions in the same group.

**Comment by securitykernel** [comment]
> Oh right, instead of `armv8crypto` ISA flag you need `armv8sha512`. For whatever reason ARM bundles the SHA-512 and SHA-3 instructions in the same group.

That is it, thanks!

**Pull request merged**